src/intrinsics/native/divmod.rs

```diff
@@ -24,219 +94,76 @@ pub fn udivmod4(
-            // r.all = n.all >> sr;
-            if sr == N_UDWORD_BITS {
-                set!(q = U256::from_words(*n.low(), 0));
-                set!(r = U256::from_words(0, *n.high()));
-            } else {
-                set!(
-                    r = U256::from_words(
-                        n.high() >> sr,
-                        (n.high() << (N_UDWORD_BITS - sr)) | (n.low() >> sr),
-                    )
-                );
-                set!(q = U256::from_words(n.low() << (N_UDWORD_BITS - sr), 0));
-            }
+            // First, divide with the high part to get the remainder in dividend.high.
+            // After that dividend.high < divisor.low.
+            *quotient.high_mut() = dividend.high() / divisor.low();
+            let dividend_high = dividend.high() % divisor.low();
+            *quotient.low_mut() = udiv256by128to128(
+                dividend_high,
+                *dividend.low(),
+                *divisor.low(),
+                remainder.low_mut(),
+            );
         }
+        if let Some(rem) = rem {
+            set!(rem = remainder);
+        }
+        set!(res = quotient);
+        return;
     }
-    // Not a special case
-    // q and r are initialized with:
-    // q.all = n.all << (N_UTWORD_BITS - sr);
-    // r.all = n.all >> sr;
-    // 1 <= sr <= N_UTWORD_BITS - 1
-    let mut carry = 0u128;
-    let mut q = unsafe { q.assume_init() };
-    let mut r = unsafe { r.assume_init() };
-    while sr > 0 {
-        // r:q = ((r:q)  << 1) | carry
-        *r.high_mut() = (r.high() << 1) | (r.low() >> (N_UDWORD_BITS - 1));
-        *r.low_mut() = (r.low() << 1) | (q.high() >> (N_UDWORD_BITS - 1));
-        *q.high_mut() = (q.high() << 1) | (q.low() >> (N_UDWORD_BITS - 1));
-        *q.low_mut() = (q.low() << 1) | carry;
-        // carry = 0;
-        // if (r.all >= d.all)
+
+    // 0 <= shift <= 127.
+    let shift = divisor.high().leading_zeros() - dividend.high().leading_zeros();
+    let mut divisor = divisor << shift;
+    let mut dividend = *dividend;
+    for _ in 0..=shift {
+        *quotient.low_mut() = quotient.low().wrapping_shl(1);
+        // Branch free version of.
+        // if (dividend >= divisor)
         // {
-        //     r.all -= d.all;
-        //      carry = 1;
+        //    dividend -= divisor;
+        //    carry = 1;
         // }
-        // NOTE: Modified from `(d - r - 1) >> (N_UTWORD_BITS - 1)` to be an
-        // **arithmetic** shift.
-        let s = {
-            let (hi, _) = d.wrapping_sub(r).wrapping_sub(U256::ONE).into_words();
-            ((hi as i128) >> 127).as_u256()
-        };
-        carry = s.low() & 1;
-        r -= d & s;
 
-        sr -= 1;
+        // const ti_int s =
+        //     (ti_int)(divisor.all - dividend.all - 1) >> (n_utword_bits - 1);
+        let mut s = divisor.wrapping_sub(dividend).wrapping_sub(U256::ONE);
+        s = U256::from_words((*s.high() as i128 >> 127) as u128, s.high() >> 127);
```

Comment:
I think the issue is this line here.

The C implementation is doing a branchless division, so basically, it computes `s` such that `s = if dividend >= divisor { 0xff...ff } else { 0 }`. However, since only the high word is being computed with arithmetic division, instead of both words, then `s = if dividend >= divisor { 0xff...ff00...01 } else { 0 }`. This makes it so only the top word (and least significant bit) of the divisor gets subtracted from the dividend.

```suggestion
        s = U256::from_words((*s.high() as i128 >> 127) as u128, (*s.high() as i128 >> 127) as u128);
```